### PR TITLE
Add instrumentation for jetty 12

### PR DESCRIPTION
**Pull request by laurit**
Resolves https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/10387

Add instrumentation and smoke tests from jetty 12.

### CODE REVIEW

File: instrumentation/jetty/jetty-12.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/jetty/v12_0/Jetty12HandlerTest.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+import java.nio.charset.StandardCharsets;
+import java.util.Collections;
+import org.eclipse.jetty.server.Handler;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.Response;
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.util.Callback;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+public class Jetty12HandlerTest extends AbstractHttpServerTest<Server> {
```
Comment (steverao) [inline instrumentation/jetty/jetty-12.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/jetty/v12_0/Jetty12HandlerTest.java:39]:
```suggestion
class Jetty12HandlerTest extends AbstractHttpServerTest<Server> {
```

File: instrumentation/jetty/jetty-12.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jetty/v12_0/Jetty12InstrumentationModule.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  }
+
+  @Override
+  public ElementMatcher.Junction<ClassLoader> classLoaderMatcher() {
+    return hasClassesNamed("org.eclipse.jetty.server.Request$Handler");
+  }
+
+  @Override
+  public List<TypeInstrumentation> typeInstrumentations() {
+    return Collections.singletonList(new Jetty12ServerInstrumentation());
```
Comment (steverao) [inline instrumentation/jetty/jetty-12.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jetty/v12_0/Jetty12InstrumentationModule.java:31]:
```suggestion
    return singletonList(new Jetty12ServerInstrumentation());
```

File: instrumentation/jetty/jetty-12.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jetty/v12_0/Jetty12Helper.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import io.opentelemetry.context.Context;
+import io.opentelemetry.instrumentation.api.instrumenter.Instrumenter;
+import io.opentelemetry.javaagent.bootstrap.servlet.AppServerBridge;
+import io.opentelemetry.javaagent.instrumentation.servlet.ServletHelper;
+import javax.annotation.Nullable;
+import org.eclipse.jetty.server.HttpStream;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.Response;
+
+public class Jetty12Helper {
```
Comment (jeanbisutti) [inline instrumentation/jetty/jetty-12.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jetty/v12_0/Jetty12Helper.java:17]:
```suggestion
class Jetty12Helper {
```
To underline that it's not the " class entry point" of the package

File: instrumentation/jetty/jetty-12.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jetty/v12_0/Jetty12HttpAttributesGetter.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+import io.opentelemetry.instrumentation.api.semconv.http.HttpServerAttributesGetter;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.util.List;
+import javax.annotation.Nullable;
+import org.eclipse.jetty.http.HttpURI;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.Response;
+
+public class Jetty12HttpAttributesGetter implements HttpServerAttributesGetter<Request, Response> {
```
Comment (jeanbisutti) [inline instrumentation/jetty/jetty-12.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jetty/v12_0/Jetty12HttpAttributesGetter.java:17]:
```suggestion
class Jetty12HttpAttributesGetter implements HttpServerAttributesGetter<Request, Response> {
```

File: instrumentation/jetty/jetty-12.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jetty/v12_0/Jetty12ResponseMutator.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.jetty.v12_0;
+
+import io.opentelemetry.javaagent.bootstrap.http.HttpServerResponseMutator;
+import org.eclipse.jetty.server.Response;
+
+public enum Jetty12ResponseMutator implements HttpServerResponseMutator<Response> {
```
Comment (jeanbisutti) [inline instrumentation/jetty/jetty-12.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jetty/v12_0/Jetty12ResponseMutator.java:11]:
```suggestion
enum Jetty12ResponseMutator implements HttpServerResponseMutator<Response> {
```

File: instrumentation/jetty/jetty-12.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jetty/v12_0/Jetty12ServerInstrumentation.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+import io.opentelemetry.javaagent.bootstrap.http.HttpServerResponseCustomizerHolder;
+import io.opentelemetry.javaagent.extension.instrumentation.TypeInstrumentation;
+import io.opentelemetry.javaagent.extension.instrumentation.TypeTransformer;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.Response;
+
+public class Jetty12ServerInstrumentation implements TypeInstrumentation {
```
Comment (jeanbisutti) [inline instrumentation/jetty/jetty-12.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jetty/v12_0/Jetty12ServerInstrumentation.java:25]:
```suggestion
class Jetty12ServerInstrumentation implements TypeInstrumentation {
```

File: instrumentation/jetty/jetty-12.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jetty/v12_0/Jetty12Singletons.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+import io.opentelemetry.instrumentation.api.semconv.http.HttpServerMetrics;
+import io.opentelemetry.instrumentation.api.semconv.http.HttpServerRoute;
+import io.opentelemetry.instrumentation.api.semconv.http.HttpSpanNameExtractor;
+import io.opentelemetry.instrumentation.api.semconv.http.HttpSpanStatusExtractor;
+import io.opentelemetry.javaagent.bootstrap.internal.CommonConfig;
+import io.opentelemetry.javaagent.bootstrap.servlet.AppServerBridge;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.Response;
+
+public final class Jetty12Singletons {
```
Comment (jeanbisutti) [inline instrumentation/jetty/jetty-12.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jetty/v12_0/Jetty12Singletons.java:23]:
```suggestion
final class Jetty12Singletons {
```

File: instrumentation/jetty/jetty-12.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jetty/v12_0/Jetty12Helper.java (hunk 3 above)
Comment (laurit) [reply instrumentation/jetty/jetty-12.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jetty/v12_0/Jetty12Helper.java:17]: As this class is used from the advice I suspect making it package private will break the advice

File: instrumentation/jetty/jetty-12.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jetty/v12_0/Jetty12Singletons.java (hunk 7 above)
Comment (laurit) [reply instrumentation/jetty/jetty-12.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jetty/v12_0/Jetty12Singletons.java:23]: as this class is used from the advice I suspect making it package private will break the advice

File: instrumentation/jetty/jetty-12.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jetty/v12_0/Jetty12ServerInstrumentation.java (hunk 6 above)
Comment (laurit) [reply instrumentation/jetty/jetty-12.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jetty/v12_0/Jetty12ServerInstrumentation.java:25]: almost all of the `TypeInstrumentation` classes we have are `public` although they could be package private

File: instrumentation/jetty/jetty-12.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jetty/v12_0/Jetty12ResponseMutator.java (hunk 5 above)
Comment (laurit) [reply instrumentation/jetty/jetty-12.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jetty/v12_0/Jetty12ResponseMutator.java]: this is also used in instrumentation and needs to be public

**Review by trask — APPROVED** [review]

**Comment by dalbani** [comment]
Thanks a lot @laurit for your work! 🙇

**Pull request merged**